### PR TITLE
Add ABCs CommutativePolynomial, MPolynomial_libsingular, InfinitePolynomial; deprecate is_Polynomial, is_MPolynomial 

### LINKED ISSUE #32709: Add ABCs CommutativePolynomial, MPolynomial_libsingular, InfinitePolynomial; deprecate is_Polynomial, is_MPolynomial
<div id="comment:0"></div>

`isinstance(x, [M]Polynomial)` replaces the use of `is_Polynomial`, `is_MPolynomial` (deprecated here).

The new class `CommutativePolynomial` is an ABC that is a common base of `Polynomial`, `MPolynomial`, and `InfinitePolynomial`. We introduce it here although there is currently no use for `isinstance(x, CommutativePolynomial)` in the library.

The new class `InfinitePolynomial` is a common base class of `InfinitePolynomial_sparse` and `InfinitePolynomial_dense`. Methods shared between both classes are moved to the new class.

The new class `MPolynomial_libsingular` is an ABCDEFIT (ABC defined exclusively for `isinstance` tests) with a unique direct subclass (see Meta-ticket #32414 and https://doc.sagemath.org/html/en/developer/packaging_sage_library.html#module-level-runtime-dependencies for this design pattern). Its purpose is to replace uses of `isinstance(x, MPolynomial_libsingular)` throughout the library. This eliminates hard run-time dependencies on Singular.

We also make `sage.rings.polynomial` a namespace package; this is needed because element implementations depend on various libraries.

CC:  @dimpase @mezzarobba @videlec @tscrim

Component: **refactoring**

Author: **Matthias Koeppe**

Reviewer: **Travis Scrimshaw**

_Issue created by migration from https://trac.sagemath.org/ticket/32709_

### PR DESCRIPTION
**Pull request by mkoeppe**
### 📚 Description

Closes #32709

### 📝 Checklist

- [ ] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [x] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mkoeppe** [comment]
PR fixes merge conflict in positively reviewed ticket #32709, setting to positive review.

**Comment by mkoeppe** [comment]
Failures in `sage/rings/polynomial/infinite_polynomial_element.py` https://github.com/sagemath/sage/actions/runs/4153778889/jobs/7185627677#step:11:7560

**Review by tscrim — APPROVED** [review]
Let it be so, but I am still a bit worried about the continued increase in maintenance burden with the lack of specifications.

**Comment by mkoeppe** [comment]
A more positive take would be that the new class `CommutativePolynomial` now is a clear place where we can put abstract methods to specify an API common to single- and multivariate polynomials.

**Comment by mkoeppe** [comment]
Thanks for the review!

**Comment by tscrim** [comment]
> A more positive take would be that the new class `CommutativePolynomial` now is a clear place where we can put abstract methods to specify an API common to single- and multivariate polynomials.

Indeed, this I am _very_ happy about (a long time wish-list item from me). However, I am still not clear what makes the infinite polynomials deserving of an ABCDEFIT.

**Comment by tscrim** [comment]
Also, could you change the PR description to "Fixes #32709" instead of "Closes #32709" for the automatic closing of the issue?

**Comment by mkoeppe** [comment]
"Fixes", "Closes", and "Resolves" all have the same magic effect.

**Comment by tscrim** [comment]
I see. Thanks.

**Comment by mkoeppe** [comment]
> > A more positive take would be that the new class `CommutativePolynomial` now is a clear place where we can put abstract methods to specify an API common to single- and multivariate polynomials.
> 
> Indeed, this I am _very_ happy about (a long time wish-list item from me).

I have opened #35149 for keeping track of follow-ups in this direction.

**Comment by tscrim** [comment]
Thank you.

File: src/sage/rings/polynomial/multi_polynomial.pyx [hunk 1]
```diff
@@ -20,6 +20,9 @@ from itertools import chain, islice
 from sage.misc.misc_c import prod
 
 def is_MPolynomial(x):
+    from sage.misc.superseded import deprecation
+    deprecation(32709, "the function is_MPolynomial is deprecated; use isinstance(x, sage.structure.element.MPolynomial) instead")
```
Comment (dimpase) [inline src/sage/rings/polynomial/multi_polynomial.pyx:24]: the deprecation message is wrong:
it should be `sage.rings.polynomial.multi_polynomial.MPolynomial`, no?

**Pull request merged**